### ISSUE TITLE
[packages] Ensure coverage of web-only Dart unit tests

### ISSUE THREAD

**Issue by stuartmorgan** [flutter/flutter#128979]
We have some tests that use `@TestOn('browser')` (or `'chrome'`, which maybe is a bug?), but in CI we only run Dart unit tests in a browser if they are part of a web plugin implementation. That means we're missing some test coverage for, e.g., `cross_file`.

We should scan for `@TestOn` in unit test folders, and if we find any, ensure that we are running tests in the appropriate mode. I believe that could just replace our current plugin-based logic, but we'll need to verify that.

**Comment by ditman** [comment]
> We have some tests that use `@TestOn('browser')` (or `'chrome'`, which maybe is a bug?)

I think we should make them all `chrome`, so when (if) we add other browsers we can land the CI changes without worrying that some random tests are going to break. Alternatively, we can make them all `browser` precisely to ensure that we can only add a new browser to CI if we guarantee that all tests are passing (your call :P).

> We should scan for `@TestOn` in unit test folders, and if we find any, ensure that we are running tests in the appropriate mode

`@TestOn` can be an arbitrarily complex string, but the [docs](https://pub.dev/packages/test#platform-selectors) contain both the format and the package that can be used to parse the string. (It's seems that `browser` result can be expressed in many different ways, like `!vm` or `chrome || firefox || safari`!)

What about trying to figure out if a package will show as "supports web" in pub.dev? If that can be computed (maybe [reusing bits of `pana`](https://github.com/dart-lang/pana/blob/master/lib/src/tag/tagger.dart#L216-L223)?), then we run all tests with `--platform=chrome` in addition to without it.

Then the `@TestOn` tags will decide what tests are actually run.

(An advantage of standardizing `@TestOn` is that the parsing argument I made above is moot, we just look for known values of the annotation, which is limited to a few known cases, and fail for unknown/unexpected values)

**Comment by stuartmorgan** [comment]
> What about trying to figure out if a package will show as "supports web" in pub.dev? If that can be computed (maybe [reusing bits of `pana`](https://github.com/dart-lang/pana/blob/master/lib/src/tag/tagger.dart#L216-L223)?), then we run all tests with `--platform=chrome` in addition to without it.
> 
> Then the `@TestOn` tags will decide what tests are actually run.

Yes, I think I was coming at this the wrong way. I was trying to improve the implementation of the "run things in the VM unless we have to run them in the browser" strategy that we had, but given that there are a bunch of subtle differences with web the better option is to bring up another CI task that runs everything in web.

That gives us maximal web coverage, and as a bonus means we don't need complex logic in the tool; we can just make a `browser` flag that enables it, and then it there are tests that don't work under that, we annotate the tests accordingly.

**Comment by ditman** [comment]
> given that there are a bunch of subtle differences with web the better option is to bring up another CI task that runs everything in web.

Ah yes, much much simpler that way, and all the pieces are already in place to allow for this, even with sharding (if needed) and exclusions.
